### DISCUSSION TITLE
Legacy Pairing

### DISCUSSION THREAD

**Discussion by pschatzmann** [pschatzmann/ESP32-A2DP#173] [General]
The basic documentation about [authentication can be found here](https://github.com/pschatzmann/ESP32-A2DP/wiki/Access-Protection---Authentication).

It is not possible to request an id from the user on the mobile phone using the compiled standard Arduino core. It is possible however when you use IDF or recompile the Arduino Core as described by Roberto Dario below:

> On 18 Feb 2022, at 17:14, Roberto Dario <robertodario@gmail.com> wrote:
> 
> Dear Phil,
> thank you for your A2DP github project!
> 
> Just a hint if I could help, about:
> https://github.com/pschatzmann/ESP32-A2DP/wiki/Access-Protection---Authentication
> 
> I get my smartphone asking for a PIN (legacy pairing) removing the Simple Secure Pairing  (SSP) that is enabled by default on Arduino Libraries.
> 
> So I recompiled the libraries removing from menuconfig the SSP, so I get on config file:
> # CONFIG_BT_SSP_ENABLED is not set
> 
> On source code I wrote:
>         esp_bt_sp_param_t param_type = ESP_BT_SP_IOCAP_MODE;
> 
>         esp_bt_io_cap_t iocap = ESP_BT_IO_CAP_IO;
> 
> //        esp_bt_gap_set_security_param(param_type, &iocap, sizeof(uint8_t));  // Missing on library when SSP is not active!
> 
>         esp_bt_pin_type_t pin_type = ESP_BT_PIN_TYPE_FIXED;
>         esp_bt_pin_code_t pin_code;
> 
>         pin_code[0]='1';
>         pin_code[1]='9';
>         pin_code[2]='7';
>         pin_code[3]='9';
>         esp_bt_gap_set_pin(pin_type, 4, pin_code);
> 
> 
> so.. when I try to pair my smartphone ask me for a PIN,
> if I type 1979 --> pairing successful
> if I type another number -> no pairing
> 
> Best Regards,
> Roberto Dario
> 

From: Roberto Dario <robertodario@gmail.com>
Subject: Re: authentication
Date: 18 February 2022 at 18:04:55 CET
To: Phil Schatzmann <phil.schatzmann@gmail.com>

No.. or better: I can explain to you what I did in my working environment (but.. maybe there are better solutions!)

I'm using Platformio with this config:

platform = https://github.com/platformio/platform-espressif32.git#feature/arduino-upstream
platform_packages = 
    platformio/framework-arduinoespressif32 @ https://github.com/espressif/arduino-esp32.git

the problem is that these libraries are compiled with SSP enabled (in particular "libbt.a")

So I downloaded v4.4 of IDF:
git clone -b v4.4 https://github.com/espressif/esp-idf.git
cd esp.idf
./install.sh
. ./export.sh

then I downloaded "arduino-lib-builder":
git clone https://github.com/espressif/esp32-arduino-lib-builder
cd esp32-arduino-lib-builder

I uncommented line #36 of build.sh (easily marked as #uncomment next line to access menuconfig), then I launched building process
./build.sh

(to be honest I also removed from line #24 useless targets for me leaving only: TARGETS="esp32")

When "menuconfig" started I unchecked:
Component Config->Bluetooth->Bluedroid Options->Secure Simple Pairing
then Save and Quit

when the build reached the end.. (over time... this point can change when libbuilder commits are not in sync with IDF!!!)
I got on "out\tools\sdk\esp32\lib" folder the new "libbt.a"

Then I copied the new library on local platformio repository:
C:\Users\Roberto.Dario\.platformio\packages\framework-arduinoespressif32\tools\sdk\esp32\lib

Cleaned my project and recompiled and PIN started working.

I don't know if this could fit your needings.. 

Best Regards,
Roberto Dario

**Comment by roschme** [comment]
Hello,

i don't get the following, can you please provide me some help?
Where i've to note the bluetooth adress of my valid device?
Thanks a lot! 

"
    You can provide a callback method which decides if a device is valid or not by calling set_address_validator()

BluetoothA2DPSink a2dp_sink;

bool check(esp_bd_addr_t adr){
  // this will prevent all connections
  return false;
}

void setup() {
  a2dp_sink.set_address_validator(check);
  a2dp_sink.start("MyMusic", false);  
}

void loop() {
}
